### DISCUSSION TITLE
How to snake_case requests with minimal api?

### DISCUSSION THREAD

**Discussion by Wuerike** [dotnet/aspnetcore#51914] [Q&A]
Hi!

I'm writing an application that requires both request paramenters and response body to be in snake_case

With the following code I've managed to serialize the response, but it doesn't affect the request deserialization.
```
builder.Services.Configure<JsonOptions>(options =>
{
    options.SerializerOptions.PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower;
});
```

Worth say that what I mean by request isn't the body, but the query parameters. At the endpoint I'm getting the query string with the `AsParameters`

I'm running 8.0.100-preview.6.23330.14

**Comment by moxplod** [comment]
Facing this exact issue today.

**Comment by davidfowl** [comment]
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/responses?view=aspnetcore-9.0#configure-json-serialization-options-globally

**Comment by moxplod** [comment]
I actually realized that my issue was the query string working differently than the global serialization options. As that uses model binding vs json serialization options. When I changed everything to snake case the document changed the request parameter to snake case as well. But it doesn’t deserialize correctly.

**Comment by seesharprun** [comment]
I have a minimal API that receives webhook events from GitHub.

If I use the configuration that @davidfowl suggested, I'm able to deserialize the body successfully into a C# POCO type:

```csharp
builder.Services.ConfigureHttpJsonOptions(options =>
{
    options.SerializerOptions.PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower;
    options.SerializerOptions.PropertyNameCaseInsensitive = true;
});
```

```csharp
app.MapPost(
    "/event-handler",
    async (
        [FromServices] IHandlerService handlerService,
        [FromBody] Payload payload
    ) => await handlerService.HandleAsync(payload));
```

My payload is just simple C# record types

```csharp
sealed record Payload
{
    public required string Action { get; init; }

    public CheckSuite? CheckSuite { get; init; }

    public PullRequest? PullRequest { get; init; }
}

sealed record CheckSuite
{
    public required long Id { get; init; }
}

internal sealed record PullRequest
{
    public required long Id { get; init; }
}
```

Ideally, I want to use a `JsonSerializerContext` so I can configure the same serialization and deserialization context for model binding, HTTP client requests, and any service that wants to manually serialize/deserialize. But that doesn't seem to work:

```csharp
[JsonSourceGenerationOptions(
    PropertyNamingPolicy = JsonKnownNamingPolicy.SnakeCaseLower,
    PropertyNameCaseInsensitive = true,
    UseStringEnumConverter = true)]
[JsonSerializable(typeof(Payload))]
internal sealed partial class GitHubJsonSerializerContext : JsonSerializerContext;
```

```csharp
// Is invalid due to the property being read-only
builder.Services.ConfigureHttpJsonOptions(
    options => options.SerializerOptions = GitHubJsonSerializerContext.Default.Options
);

builder.Services.Configure<JsonSerializerOptions>(
    static o => o.TypeInfoResolver = GitHubJsonSerializerContext.Default
);
```

Am I forgetting something obvious? The `SerializerOptions` property is read-only, so I can only set a single property at a time. This is as far as I got:

```csharp
builder.Services.ConfigureHttpJsonOptions(
    static options => options.SerializerOptions.TypeInfoResolver = GitHubJsonSerializerContext.Default
);
```

Unfortunately this leaves the naming convention to `JsonNamingPolicy.CamelCase`. ☹️

**Reply by martincostello** [reply]
This doesn't answer your question directly, but have you considered using the [Octokit.Webhooks.AspNetCore](https://www.nuget.org/packages/Octokit.Webhooks.AspNetCore) NuGet package?

This provides an endpoint to receive the webhooks, signature validation of the payloads if needed, comes with all the POCOs you need and handles all the serialization for you.

**Reply by seesharprun** [reply]
I will investigate that package. I was building it all out manually. Thankfully, building it out manually was a useful exercise to help me understand how each webhook works internally :-).

Thanks!
